src/Storages/KeyDescriptionMonotonicityChecker.cpp

```diff
@@ -0,0 +1,249 @@
+        return (*col)[0];
+    }
+
+    FieldRef applyFunction(const FunctionBasePtr & func, const DataTypePtr & current_type, const FieldRef & field)
+    {
+        /// Fallback for fields without block reference.
+        if (field.isExplicit())
+            return applyFunctionForField(func, current_type, field);
+
+        String result_name = "_" + func->getName() + "_" + toString(field.column_idx);
+        const auto & columns = field.columns;
+        size_t result_idx = columns->size();
+
+        for (size_t i = 0; i < result_idx; ++i)
+        {
+            if ((*columns)[i].name == result_name)
+                result_idx = i;
+        }
+
+        if (result_idx == columns->size())
+        {
+            ColumnsWithTypeAndName args{(*columns)[field.column_idx]};
+            field.columns->emplace_back(ColumnWithTypeAndName {nullptr, func->getResultType(), result_name});
+            (*columns)[result_idx].column = func->execute(args, (*columns)[result_idx].type, columns->front().column->size());
+        }
+
+        return {field.columns, field.row_idx, result_idx};
+    }
+
+    IFunction::Monotonicity applyMonotonicFunctionsChainToRange(
+        Range key_range,
+        const std::vector<FunctionBasePtr> & functions,
+        DataTypePtr current_type)
+    {
+        // default to monotonically increasing, satisfies cases where function list is empty
+        IFunction::Monotonicity monotonicity {true, true, true, true};
+
+        for (const auto & func : functions)
+        {
+            /// We check the monotonicity of each function on a specific range.
+            /// If we know the given range only contains one value, then we treat all functions as positive monotonic.
+            monotonicity = func->getMonotonicityForRange(*current_type.get(), key_range.left, key_range.right);
+
+            if (!monotonicity.is_monotonic)
+            {
+                return {};
+            }
+
+            /// If we apply function to open interval, we can get empty intervals in result.
+            /// E.g. for ('2020-01-03', '2020-01-20') after applying 'toYYYYMM' we will get ('202001', '202001').
+            /// To avoid this we make range left and right included.
+            /// Any function that treats NULL specially is not monotonic.
+            /// Thus we can safely use isNull() as an -Inf/+Inf indicator here.
+            if (!key_range.left.isNull())
+            {
+                key_range.left = applyFunction(func, current_type, key_range.left);
+                key_range.left_included = true;
+            }
+
+            if (!key_range.right.isNull())
+            {
+                key_range.right = applyFunction(func, current_type, key_range.right);
+                key_range.right_included = true;
+            }
+
+            current_type = func->getResultType();
+
+            if (!monotonicity.is_positive)
+                key_range.invert();
```

Comment:
Let's avoid copy-paste if possible

Reply:
Ok, I believe I can make use of the `KeyCondition` version with some small adaptations.